src/ucp/core/ucp_ep.c

```diff
@@ -3415,6 +3416,73 @@ void ucp_ep_reqs_purge(ucp_ep_h ucp_ep, ucs_status_t status)
     }
 }
 
+static ucs_status_t ucp_ep_query_transport(ucp_ep_h ep, ucp_ep_attr_t *attr)
+{
+    ucp_worker_h worker     = ep->worker;
+    ucp_ep_config_t *config = ucp_ep_config(ep);
+    const uct_tl_resource_desc_t *rsc;
+    ucp_transport_entry_t *transport_entry;
+    size_t device_limit;
+    size_t transport_limit;
+    ucp_lane_index_t lane_index;
+    ucp_rsc_index_t cm_idx;
+    ucp_rsc_index_t rsc_index;
+
+    /* Compute field end offsets for each field in the ucp_transport_entry_t
+     * structure. */
+    transport_limit = ucs_offsetof(ucp_transport_entry_t, transport_name) +
+                      sizeof(transport_entry->transport_name);
+    device_limit    = ucs_offsetof(ucp_transport_entry_t, device_name) +
+                      sizeof(transport_entry->device_name);
+
+    for (lane_index = 0; lane_index < ucs_min(attr->transports.num_entries,
+                                              config->key.num_lanes);
+         lane_index++) {
+        /* Since the caller may be using a different size ucp_transport_entry_t
+         * structure definiition than this code, array indexing cannot be used
+         * when accesing array elements. The array element's offset must be computed
+         * as 'lane_index' * attr->transports.entry_size and that offset added to the
+         * array's base address. */
+        transport_entry =
+                UCS_PTR_BYTE_OFFSET(attr->transports.entries,
+                                    lane_index * attr->transports.entry_size);
+
+        /* Each field updated in the following block must have its ending offset
+         * compared to attr->transports.entry_size before the field is 
+         * updated. If the field's ending offset is greater than the 
+         * attr->transports.entry_size value, the field cannot be updated because
+         * that will cause a storage overlay.
+         */
+        if (lane_index == config->key.cm_lane) {
+            cm_idx = ucp_ep_ext_control(ep)->cm_idx;
+            if (transport_limit <= attr->transports.entry_size) {
```

Comment:
seems like this approach may lead to uninitialized fields or user has to initialize the structure with some magic values prior the call to understand what's filled, so maybe `field_mask`?

otherwise, can `entry_size` be handled once in the beginning like parameters check?

Reply:
I don't understand what you mean by handling entry_size once in the beginning. I think that with either checking entry_size or checking a field mask has to happen inside the loop where I have to make a check before I update a field. The only other way I think I can do this is a separate loop for each field where I check entry_size before running the loop, but that feels like a worse solution. I think both entry_size and field_mask have similar risks for uninitialized values as well.

Reply:
`field_mask` could be in/out parameter... but you still need a structure size for indexing, so maybe add something like `initialized_field_mask`?

Reply:
@evgeny-leksikov if we already have struct size parameter, IMO adding also a field_mask would be a redundancy and extra complication, we discussed it in the API PR https://github.com/openucx/ucx/pull/7990.
so the expectation here is that user should initialize all fields (up to the provided structure size)

Reply:
My point is `transport_name` and `device_name` are out-parameters which should be initialized by UCP. If the app is compiled with newer UCP version with `new_field` then `entry_size` works only for correct indexing in old UCP version, old UCP is not able to initialize `new_field` and no way to report it to app.

Reply:
ok then